SQL error when logging in for first time (per user) after Keycloak upgrade

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

After upgrading from an older version to Keycloak 26.2.5, there is a step where the password is upgraded from the past format to argon2. 

When the user tries to login they get an "Internal server error" message and in the Keycloak logs I see:
Caused by: org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update CREDENTIAL set CREATED_DATE=?,CREDENTIAL_DATA=?,PRIORITY=?,SALT=?,SECRET_DATA=?,TYPE=?,USER_ID=?,USER_LABEL=?,VERSION=? where ID=? and VERSION=?

If they retry the login again, they're logged in and the password is upgraded.

### Version

26.2.5

### Regression

- [x] The issue is a regression

### Expected behavior

I would expect the user to be logged in and the password upgrade to happen seamlessly. 

### Actual behavior

The first login attempt fails, the second login attempt succeeds. 

### How to Reproduce?

Upgrade from a much older version of Keycloak (using stepped upgrades between major changes like the move from Wildfly to Quarkus and then first Quarkus to latest Quarkus), and try to login. 

### Anything else?

_No response_

On the browser side, they see an "Internal server error" message and a failed login.

On the server side, here's the full stack trace:

`2025-06-03T01:26:23.216915+00:00 HOSTNAME kc.sh[3500]: 2025-06-03 01:26:23,215 INFO  [org.hibernate.orm.jdbc.batch] (executor-thread-72) HHH100503: On release of batch it still contained JDBC statements
2025-06-03T01:26:23.221599+00:00 HOSTNAME kc.sh[3500]: 2025-06-03 01:26:23,218 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-72) Uncaught server error: org.keycloak.models.ModelException: Database operation failed
2025-06-03T01:26:23.221721+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.connections.jpa.PersistenceExceptionConverter.convert(PersistenceExceptionConverter.java:106)
2025-06-03T01:26:23.221788+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.connections.jpa.JpaExceptionConverter.convert(JpaExceptionConverter.java:31)
2025-06-03T01:26:23.221839+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.transaction.JtaTransactionWrapper.lambda$handleException$0(JtaTransactionWrapper.java:77)
2025-06-03T01:26:23.221872+00:00 HOSTNAME kc.sh[3500]: #011at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2025-06-03T01:26:23.221903+00:00 HOSTNAME kc.sh[3500]: #011at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1787)
2025-06-03T01:26:23.221941+00:00 HOSTNAME kc.sh[3500]: #011at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
2025-06-03T01:26:23.221976+00:00 HOSTNAME kc.sh[3500]: #011at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
2025-06-03T01:26:23.222011+00:00 HOSTNAME kc.sh[3500]: #011at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
2025-06-03T01:26:23.222046+00:00 HOSTNAME kc.sh[3500]: #011at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
2025-06-03T01:26:23.222082+00:00 HOSTNAME kc.sh[3500]: #011at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2025-06-03T01:26:23.222127+00:00 HOSTNAME kc.sh[3500]: #011at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
2025-06-03T01:26:23.222165+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.transaction.JtaTransactionWrapper.handleException(JtaTransactionWrapper.java:79)
2025-06-03T01:26:23.222200+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.transaction.JtaTransactionWrapper.commit(JtaTransactionWrapper.java:104)
2025-06-03T01:26:23.222234+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.services.DefaultKeycloakTransactionManager.lambda$commitWithTracing$0(DefaultKeycloakTransactionManager.java:169)
2025-06-03T01:26:23.222271+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.tracing.NoopTracingProvider.trace(NoopTracingProvider.java:59)
2025-06-03T01:26:23.222308+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.tracing.NoopTracingProvider.trace(NoopTracingProvider.java:69)
2025-06-03T01:26:23.222343+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.services.DefaultKeycloakTransactionManager.commitWithTracing(DefaultKeycloakTransactionManager.java:168)
2025-06-03T01:26:23.222380+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:136)
2025-06-03T01:26:23.222420+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.services.DefaultKeycloakSession.closeTransactionManager(DefaultKeycloakSession.java:396)
2025-06-03T01:26:23.222473+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.services.DefaultKeycloakSession.close(DefaultKeycloakSession.java:361)
2025-06-03T01:26:23.222510+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.models.KeycloakBeanProducer_ProducerMethod_getKeycloakSession_XoSEUTXOsE3bpqXlGMAykCiECUM_ClientProxy.close(Unknown Source)
2025-06-03T01:26:23.222559+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.quarkus.runtime.transaction.TransactionalSessionHandler.close(TransactionalSessionHandler.java:60)
2025-06-03T01:26:23.222603+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.quarkus.runtime.integration.jaxrs.CloseSessionFilter.closeSession(CloseSessionFilter.java:67)
2025-06-03T01:26:23.222636+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.quarkus.runtime.integration.jaxrs.CloseSessionFilter.filter(CloseSessionFilter.java:63)
2025-06-03T01:26:23.222669+00:00 HOSTNAME kc.sh[3500]: #011at org.jboss.resteasy.reactive.server.handlers.ResourceResponseFilterHandler.handle(ResourceResponseFilterHandler.java:25)
2025-06-03T01:26:23.222705+00:00 HOSTNAME kc.sh[3500]: #011at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:150)
2025-06-03T01:26:23.222740+00:00 HOSTNAME kc.sh[3500]: #011at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
2025-06-03T01:26:23.222778+00:00 HOSTNAME kc.sh[3500]: #011at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:638)
2025-06-03T01:26:23.222813+00:00 HOSTNAME kc.sh[3500]: #011at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
2025-06-03T01:26:23.222850+00:00 HOSTNAME kc.sh[3500]: #011at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
2025-06-03T01:26:23.222884+00:00 HOSTNAME kc.sh[3500]: #011at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
2025-06-03T01:26:23.222918+00:00 HOSTNAME kc.sh[3500]: #011at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
2025-06-03T01:26:23.222954+00:00 HOSTNAME kc.sh[3500]: #011at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
2025-06-03T01:26:23.222985+00:00 HOSTNAME kc.sh[3500]: #011at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
2025-06-03T01:26:23.223018+00:00 HOSTNAME kc.sh[3500]: #011at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2025-06-03T01:26:23.223053+00:00 HOSTNAME kc.sh[3500]: #011at java.base/java.lang.Thread.run(Thread.java:1583)
2025-06-03T01:26:23.223091+00:00 HOSTNAME kc.sh[3500]: Caused by: org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update CREDENTIAL set CREATED_DATE=?,CREDENTIAL_DATA=?,PRIORITY=?,SALT=?,SECRET_DATA=?,TYPE=?,USER_ID=?,USER_LABEL=?,VERSION=? where ID=? and VERSION=?
2025-06-03T01:26:23.223244+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.jdbc.Expectations.checkBatched(Expectations.java:80)
2025-06-03T01:26:23.223291+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.jdbc.Expectation$RowCount.verifyOutcome(Expectation.java:174)
2025-06-03T01:26:23.223327+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.engine.jdbc.batch.internal.BatchImpl.checkRowCounts(BatchImpl.java:319)
2025-06-03T01:26:23.223360+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.engine.jdbc.batch.internal.BatchImpl.lambda$performExecution$2(BatchImpl.java:285)
2025-06-03T01:26:23.223396+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.engine.jdbc.mutation.internal.PreparedStatementGroupSingleTable.forEachStatement(PreparedStatementGroupSingleTable.java:67)
2025-06-03T01:26:23.223435+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.engine.jdbc.batch.internal.BatchImpl.performExecution(BatchImpl.java:264)
2025-06-03T01:26:23.223472+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.engine.jdbc.batch.internal.BatchImpl.execute(BatchImpl.java:242)
2025-06-03T01:26:23.223508+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.executeBatch(JdbcCoordinatorImpl.java:188)
2025-06-03T01:26:23.223874+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:674)
2025-06-03T01:26:23.223933+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:511)
2025-06-03T01:26:23.223985+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:414)
2025-06-03T01:26:23.224046+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:41)
2025-06-03T01:26:23.224100+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
2025-06-03T01:26:23.224152+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1429)
2025-06-03T01:26:23.224201+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:491)
2025-06-03T01:26:23.224247+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2354)
2025-06-03T01:26:23.224299+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:1978)
2025-06-03T01:26:23.224349+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:439)
2025-06-03T01:26:23.224401+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:336)
2025-06-03T01:26:23.224470+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
2025-06-03T01:26:23.224546+00:00 HOSTNAME kc.sh[3500]: #011at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
2025-06-03T01:26:23.225418+00:00 HOSTNAME kc.sh[3500]: #011at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:52)
2025-06-03T01:26:23.225568+00:00 HOSTNAME kc.sh[3500]: #011at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:348)
2025-06-03T01:26:23.225637+00:00 HOSTNAME kc.sh[3500]: #011at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:66)
2025-06-03T01:26:23.225708+00:00 HOSTNAME kc.sh[3500]: #011at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:135)
2025-06-03T01:26:23.225769+00:00 HOSTNAME kc.sh[3500]: #011at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1307)
2025-06-03T01:26:23.225837+00:00 HOSTNAME kc.sh[3500]: #011at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:104)
2025-06-03T01:26:23.225907+00:00 HOSTNAME kc.sh[3500]: #011at io.quarkus.narayana.jta.runtime.NotifyingTransactionManager.commit(NotifyingTransactionManager.java:70)
2025-06-03T01:26:23.226058+00:00 HOSTNAME kc.sh[3500]: #011at org.keycloak.transaction.JtaTransactionWrapper.commit(JtaTransactionWrapper.java:102)
2025-06-03T01:26:23.226128+00:00 HOSTNAME kc.sh[3500]: #011... 23 more
`

--- a/services/src/main/java/org/keycloak/credential/PasswordCredentialProvider.java
+++ b/services/src/main/java/org/keycloak/credential/PasswordCredentialProvider.java
@@ -22,6 +22,7 @@ import io.micrometer.core.instrument.Tag;
 import org.jboss.logging.Logger;
 import org.keycloak.common.util.Time;
 import org.keycloak.credential.hash.PasswordHashProvider;
+import org.keycloak.models.AbstractKeycloakTransaction;
 import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.ModelException;
 import org.keycloak.models.PasswordPolicy;
@@ -282,14 +283,26 @@ public class PasswordCredentialProvider implements CredentialProvider<PasswordCr
         if (!provider.policyCheck(passwordPolicy, password)) {
             final int iterations = passwordPolicy != null ? passwordPolicy.getHashIterations() : -1;
             final String hashAlgorithm = passwordPolicy != null ? passwordPolicy.getHashAlgorithm() : null;
-            try {
-                // refresh the password in a different transaction, do not fail if there is a model exception
-                KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), session.getContext(),
-                        (KeycloakSession s) -> refreshPassword(s, hashAlgorithm, iterations, input.getChallengeResponse(),
-                                password.getId(), password.getCreatedDate(), password.getUserLabel(), user.getId()));
-            } catch (ModelException e) {
-                logger.info("Error re-hashing the password in a different transaction", e);
-            }
+            // Refresh the password in a different transaction, do not fail if there is a model exception on current modifications due to concurrent logins.
+            // Also do not start it as a nested transaction, as the current transaction might have auto-migrated the credential.
+            // see: JpaUserCredentialStore#toModel for the on-the-fly migration of the salt column
+            session.getTransactionManager().enlistAfterCompletion(new AbstractKeycloakTransaction() {
+                @Override
+                protected void commitImpl() {
+                    try {
+                        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), session.getContext(),
+                                (KeycloakSession s) -> refreshPassword(s, hashAlgorithm, iterations, input.getChallengeResponse(),
+                                        password.getId(), password.getCreatedDate(), password.getUserLabel(), user.getId()));
+                    } catch (ModelException e) {
+                        logger.info("Error re-hashing the password in a different transaction", e);
+                    }
+                }
+
+                @Override
+                protected void rollbackImpl() {
+
+                }
+            });
         }
     }
 

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/PasswordHashingTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/PasswordHashingTest.java
@@ -16,6 +16,7 @@
  */
 package org.keycloak.testsuite.forms;
 
+import jakarta.persistence.EntityManager;
 import jakarta.ws.rs.BadRequestException;
 import org.bouncycastle.crypto.generators.Argon2BytesGenerator;
 import org.jboss.arquillian.graphene.page.Page;
@@ -24,6 +25,7 @@ import org.junit.Assume;
 import org.junit.Test;
 import org.keycloak.common.crypto.FipsMode;
 import org.keycloak.common.util.Base64;
+import org.keycloak.connections.jpa.JpaConnectionProvider;
 import org.keycloak.credential.CredentialModel;
 import org.keycloak.credential.hash.PasswordHashProvider;
 import org.keycloak.credential.hash.PasswordHashProviderFactory;
@@ -35,8 +37,10 @@ import org.keycloak.crypto.hash.Argon2Parameters;
 import org.keycloak.crypto.hash.Argon2PasswordHashProviderFactory;
 import org.keycloak.models.RealmModel;
 import org.keycloak.models.UserModel;
+import org.keycloak.models.cache.UserCache;
 import org.keycloak.models.credential.PasswordCredentialModel;
 import org.keycloak.models.credential.dto.PasswordCredentialData;
+import org.keycloak.models.jpa.entities.CredentialEntity;
 import org.keycloak.representations.idm.CredentialRepresentation;
 import org.keycloak.representations.idm.ErrorRepresentation;
 import org.keycloak.representations.idm.RealmRepresentation;
@@ -52,6 +56,7 @@ import org.keycloak.testsuite.util.UserBuilder;
 
 import javax.crypto.SecretKeyFactory;
 import javax.crypto.spec.PBEKeySpec;
+import java.nio.charset.StandardCharsets;
 import java.security.spec.KeySpec;
 import java.time.Duration;
 import java.util.List;
@@ -109,6 +114,42 @@ public class PasswordHashingTest extends AbstractTestRealmKeycloakTest {
 
         loginPage.open();
         loginPage.login(username, password);
+        appPage.assertCurrent();
+
+        credential = PasswordCredentialModel.createFromCredentialModel(fetchCredentials(username));
+
+        assertEquals(Pbkdf2PasswordHashProviderFactory.ID, credential.getPasswordCredentialData().getAlgorithm());
+        assertEncoded(credential, password, credential.getPasswordSecretData().getSalt(), "PBKDF2WithHmacSHA1", 1);
+    }
+
+    @Test
+    public void testPasswordRehashedOnAlgorithmChangedWithMigratedSalt() throws Exception {
+        setPasswordPolicy("hashAlgorithm(" + Pbkdf2Sha256PasswordHashProviderFactory.ID + ") and hashIterations(1)");
+
+        String username = "testPasswordRehashedOnAlgorithmChangedWithMigratedSalt";
+        final String password = createUser(username);
+
+        PasswordCredentialModel credential = PasswordCredentialModel.createFromCredentialModel(fetchCredentials(username));
+
+        assertEquals(Pbkdf2Sha256PasswordHashProviderFactory.ID, credential.getPasswordCredentialData().getAlgorithm());
+
+        assertEncoded(credential, password, credential.getPasswordSecretData().getSalt(), "PBKDF2WithHmacSHA256", 1);
+
+        setPasswordPolicy("hashAlgorithm(" + Pbkdf2PasswordHashProviderFactory.ID + ") and hashIterations(1)");
+
+        String credentialId = credential.getId();
+        testingClient.server().run(session -> {
+            EntityManager em = session.getProvider(JpaConnectionProvider.class).getEntityManager();
+            CredentialEntity credentialEntity = em.find(CredentialEntity.class, credentialId);
+            // adding a dummy value to the salt column to trigger migration in JpaUserCredentialStore#toModel on next fetch of the credential
+            credentialEntity.setSalt("dummy".getBytes(StandardCharsets.UTF_8));
+            // Clearing the user cache as we updated the database directly
+            session.getProvider(UserCache.class).clear();
+        });
+
+        loginPage.open();
+        loginPage.login(username, password);
+        appPage.assertCurrent();
 
         credential = PasswordCredentialModel.createFromCredentialModel(fetchCredentials(username));
 
@@ -131,6 +172,7 @@ public class PasswordHashingTest extends AbstractTestRealmKeycloakTest {
 
         loginPage.open();
         loginPage.login(username, password);
+        appPage.assertCurrent();
 
         credential = PasswordCredentialModel.createFromCredentialModel(fetchCredentials(username));
 
@@ -152,6 +194,7 @@ public class PasswordHashingTest extends AbstractTestRealmKeycloakTest {
 
         loginPage.open();
         loginPage.login(username, password);
+        appPage.assertCurrent();
 
         credential = PasswordCredentialModel.createFromCredentialModel(fetchCredentials(username));
 
@@ -180,6 +223,7 @@ public class PasswordHashingTest extends AbstractTestRealmKeycloakTest {
 
         loginPage.open();
         loginPage.login(username, password);
+        appPage.assertCurrent();
 
         credential = PasswordCredentialModel.createFromCredentialModel(fetchCredentials(username));
 
@@ -192,6 +236,7 @@ public class PasswordHashingTest extends AbstractTestRealmKeycloakTest {
 
         loginPage.open();
         loginPage.login(username, password);
+        appPage.assertCurrent();
 
         credential = PasswordCredentialModel.createFromCredentialModel(fetchCredentials(username));
 

